### PR TITLE
add codeshell example

### PR DESCRIPTION
**Pull request by MeouSker77**
## Description

Add codeshell example

# CodeShell

In this directory, you'll find how to use this codeshell server with vscode codeshell extension.

## 0. Extra Environment Preparation

Suppose you have already configured GPU environment, you will need some extra preparation

1. install extra requirements
    ```
    pip install uvicorn fastapi sse_starlette
    ```

2. search `codeshell` in vscode extension market, then install `CodeShell VSCode Extension` extension

3. change extension settings:
    - change `Code Shell: Run Env For LLMs` to `GPU with TGI toolkit`
    - disable `Code Shell: Auto Trigger Completion` (use `Alt + \` to trigger completion manually)

4. download CodeShell-7B-Chat (don't use CodeShell-7B)

## 1. How to use this server

python server.py [--option value]

1. `--checkpoint-path <path>`: path to huggingface model checkpoint
2. `--device xpu`: enable GPU or not
3. `--multi-turn`: enable multi turn conversation or just support single turn conversation
4. `--cpu-embedding`: move Embedding layer to CPU or not
5. `--max-context <number>`: Clip the context length in Code Completion, it won't affect other features, set it to 99999 to disable it

## 2. Note

In my test, if use vscode remote connection to connect to a remote machine, then install extension and running this server on that remote machine, all extension features expect for Code Completion can be used.

If don't use remote conection, then install extension and running this server on local machine, Code Completion can also be used.

### CODE REVIEW

File: python/llm/example/GPU/HF-Transformers-AutoModels/Model/codeshell/README.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
+# CodeShell
+
+In this directory, you'll find how to use this codeshell server with vscode codeshell extension.
+
+## 0. Extra Environment Preparation
+
+Suppose you have already configured GPU environment, you will need some extra preparation
```
Comment (rnwang04) [inline python/llm/example/GPU/HF-Transformers-AutoModels/Model/codeshell/README.md:7]: preparations

File: python/llm/example/GPU/HF-Transformers-AutoModels/Model/codeshell/README.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    pip install uvicorn fastapi sse_starlette
+    ```
+
+2. search `codeshell` in vscode extension market, then install `CodeShell VSCode Extension` extension
+
+3. change extension settings:
+    - change `Code Shell: Run Env For LLMs` to `GPU with TGI toolkit`
+    - disable `Code Shell: Auto Trigger Completion` (use `Alt + \` to trigger completion manually)
+
+4. download CodeShell-7B-Chat (don't use CodeShell-7B)
```
Comment (rnwang04) [inline python/llm/example/GPU/HF-Transformers-AutoModels/Model/codeshell/README.md:20]: [WisdomShell/CodeShell-7B-Chat](https://huggingface.co/WisdomShell/CodeShell-7B-Chat)

File: python/llm/example/GPU/HF-Transformers-AutoModels/Model/codeshell/README.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+2. `--device xpu`: enable GPU or not
+3. `--multi-turn`: enable multi turn conversation or just support single turn conversation
+4. `--cpu-embedding`: move Embedding layer to CPU or not
+5. `--max-context <number>`: Clip the context length in Code Completion, it won't affect other features, set it to 99999 to disable it
+
+## 2. Note
+
+In my test, if use vscode remote connection to connect to a remote machine, then install extension and running this server on that remote machine, all extension features expect for Code Completion can be used.
+
+If don't use remote conection, then install extension and running this server on local machine, Code Completion can also be used.
```
Comment (rnwang04) [inline python/llm/example/GPU/HF-Transformers-AutoModels/Model/codeshell/README.md:38]: a little confusing...
You mean vscode remote connection and remote desktop connection ?

File: python/llm/example/GPU/HF-Transformers-AutoModels/Model/codeshell/server.py [hunk 4]
```diff
@@ -0,0 +1,302 @@
+# coding=utf-8
+# Copyright 2023 WisdomShell Inc. All Rights Reserved.
+
```
Comment (rnwang04) [inline python/llm/example/GPU/HF-Transformers-AutoModels/Model/codeshell/server.py:3]: No ` Copyright 2016 The BigDL Authors.` ?

File: python/llm/example/GPU/HF-Transformers-AutoModels/Model/codeshell/server.py [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+        """Returns true if all generated sequences contain any of the end-of-function strings."""
+        texts =  [ self.tokenizer.decode(ids[self.input_length:]) for ids in input_ids ]
+        dones = [ any(stop_word in text for stop_word in self.stop_words) for text in texts ]
+        # print('-'*80)
+        # print(texts)
+        # print(dones)
+        # print('output length:', len(input_ids[0]) - self.input_length)
+        # print('total length:', len(input_ids[0]))
+        # print('dones:', dones)
+        # print('-'*80)
```
Comment (rnwang04) [inline python/llm/example/GPU/HF-Transformers-AutoModels/Model/codeshell/server.py:87]: maybe just remove them.

**Review by rnwang04 — APPROVED** [review]
others LGTM

File: python/llm/example/GPU/HF-Transformers-AutoModels/Model/codeshell/README.md (hunk 3 above)
Comment (MeouSker77) [reply python/llm/example/GPU/HF-Transformers-AutoModels/Model/codeshell/README.md:38]: yes

**Pull request merged**